File: repos/SBS-Football-Drafts-main/models/draft-token.go
package models

import (
	"fmt"

	"github.com/Spoiled-Banana-Society/SBS-Football-Drafts/utils"
)

type Prizes struct {
	ETH float64 `json:"ETH"`
}

type DraftToken struct {
	Roster            *TokenRoster `json:"roster"`
	DraftType         string       `json:"_draftType"`
	CardId            string       `json:"_cardId"`
	ImageUrl          string       `json:"_imageUrl"`
	Level             string       `json:"_level"`
	OwnerId           string       `json:"_ownerId"`
	LeagueId          string       `json:"_leagueId"`
	LeagueDisplayName string       `json:"_leagueDisplayName"`
	Rank              string       `json:"_rank"`
	LeagueRank        string       `json:"_leagueRank"`
	WeekScore         string       `json:"_weekScore"`
	SeasonScore       string       `json:"_seasonScore"`
	Prizes            Prizes       `json:"prizes"`
}

type Metadata struct {
	Description string          `json:"description"`
	Name        string          `json:"name"`
	Image       string          `json:"image"`
	Attributes  []AttributeType `json:"attributes"`
}

type AttributeType struct {
	Trait_Type string `json:"trait_type"`
	Value      string `json:"value"`
}

func (dt *DraftToken) ConvertToMetadata() *Metadata {
	return &Metadata{
		Description: "Banana Best Ball, the first ever Web3 Fantasy Football Draft tournament on chain.",
		Name:        fmt.Sprintf("BBB pass #%s", dt.CardId),
		Image:       dt.ImageUrl,
		Attributes:  CreateTokenAttributes(dt),
	}
}

func CreateTokenAttributes(dt *DraftToken) []AttributeType {
	res := make([]AttributeType, 0)
	for i := 0; i < len(dt.Roster.QB); i++ {
		obj := AttributeType{
			Trait_Type: fmt.Sprintf("QB%d", i+1),
			Value:      dt.Roster.QB[i].DisplayName,
		}
		res = append(res, obj)
	}
	for i := 0; i < len(dt.Roster.RB); i++ {
		obj := AttributeType{
			Trait_Type: fmt.Sprintf("RB%d", i+1),
			Value:      dt.Roster.RB[i].DisplayName,
		}
		res = append(res, obj)
	}
	for i := 0; i < len(dt.Roster.TE); i++ {
		obj := AttributeType{
			Trait_Type: fmt.Sprintf("TE%d", i+1),
			Value:      dt.Roster.TE[i].DisplayName,
		}
		res = append(res, obj)
	}
	for i := 0; i < len(dt.Roster.WR); i++ {
		obj := AttributeType{
			Trait_Type: fmt.Sprintf("WR%d", i+1),
			Value:      dt.Roster.WR[i].DisplayName,
		}
		res = append(res, obj)
	}
	for i := 0; i < len(dt.Roster.DST); i++ {
		obj := AttributeType{
			Trait_Type: fmt.Sprintf("DST%d", i+1),
			Value:      dt.Roster.DST[i].DisplayName,
		}
		res = append(res, obj)
	}

	levelTrait := AttributeType{
		Trait_Type: "LEVEL",
		Value:      dt.Level,
	}
	res = append(res, levelTrait)
	fmt.Println("added level attribute")

	weekScoreTrait := AttributeType{
		Trait_Type: "WEEK-SCORE",
		Value:      dt.WeekScore,
	}
	res = append(res, weekScoreTrait)
	fmt.Println("added week score attribute")

	seasonScoreTrait := AttributeType{
		Trait_Type: "SEASON-SC0RE",
		Value:      dt.SeasonScore,
	}
	res = append(res, seasonScoreTrait)
	fmt.Println("added season score attribute")

	rankTrait := AttributeType{
		Trait_Type: "RANK",
		Value:      dt.Rank,
	}
	res = append(res, rankTrait)
	fmt.Println("added rank attribute")

	leagueTrait := AttributeType{
		Trait_Type: "LEAGUE-NAME",
		Value:      dt.LeagueDisplayName,
	}
	res = append(res, leagueTrait)
	fmt.Println("added league name attribute")

	leagueRankTrait := AttributeType{
		Trait_Type: "LEAGUE-RANK",
		Value:      dt.LeagueRank,
	}
	res = append(res, leagueRankTrait)
	fmt.Println("added league rank attribute")

	prizesTrait := AttributeType{
		Trait_Type: "PRIZES",
		Value:      fmt.Sprintf("%f ETH", dt.Prizes.ETH),
	}
	res = append(res, prizesTrait)
	fmt.Println("added prizes attribute")

	return res
}

func (token *DraftToken) GetDraftTokenFromDraftById(tokenId, draftId string) error {
	err := utils.Db.ReadDocument(fmt.Sprintf("drafts/%s/cards", draftId), tokenId, token)
	if err != nil {
		return err
	}
	return nil
}

File: repos/SBS-Backend-main/functions/services/draftToken.js
//PACKAGES
const web3 = require('web3');
const { v4: uuidv4 } = require('uuid');
const { FieldValue } = require('firebase-admin/firestore');

//SERVICES
const env = require('./env');
const db = require('./db');
const api = require('./api');
const sbs = require('./sbs');
const utils = require('./utils');

const internals = {};

internals.UpdateDraftToken = async (draftToken) => {
    if (draftToken.Roster.QB.isArray() == false) {
        return "invalid roster"
    } else if (draftToken.Roster.RB.isArray() == false) {
        return "invalid roster"
    } else if (draftToken.Roster.TE.isArray() == false) {
        return "invalid roster"
    } else if (draftToken.Roster.WR.isArray() == false) {
        return "invalid roster"
    } else if (draftToken.Roster.DST.isArray() == false) {
        return "invalid roster"
    } else if (typeof draftToken._draftType != "string") {
        return "draft type is not a string"
    } else if (typeof draftToken._cardId != "string") {
        return "draft type is not a string"
    } else if (typeof draftToken._imageUrl != "string") {
        return "draft type is not a string"
    } else if (typeof draftToken._level != "string") {
        return "draft type is not a string"
    } else if (typeof draftToken._ownerId != "string") {
        return "draft type is not a string"
    } else if (typeof draftToken._leagueId != "string") {
        return "draft type is not a string"
    } else if (typeof draftToken._rank != "string") {
        return "draft type is not a string"
    } else if (typeof draftToken._weekScore != "string") {
        return "draft type is not a string"
    } else if (typeof draftToken._seasonScore != "string") {
        return "draft type is not a string"
    }

    let res = await db.createOrUpdateDocument(`draftTokens`, draftToken._cardId, draftToken)
    res.isEqual



}


// internals.CheckDraftTokenOwnership = async () => {
//     const cardIds = await db.readAllDocumentIds('draftTokens')
//     for(let i = 0; i < cardIds.length; i++) {
//         const cardId = cardIds[i];
//         const card = await db.readDocument('')
//     }
// }